src/v/kafka/server/replicated_partition.cc

```diff
@@ -205,4 +207,46 @@ std::optional<model::offset> replicated_partition::get_leader_epoch_last_offset(
     }
     return std::nullopt;
 }
+
+ss::future<bool> replicated_partition::is_fetch_offset_valid(
+  model::offset fetch_offset, model::timeout_clock::time_point deadline) {
+    /**
+     * Make sure that we will update high watermark offset if it isn't yet
+     * initialized.
+     *
+     * We use simple heuristic here to determine if high watermark update is
+     * required.
+     *
+     * We only request update if fetch_offset is in range beteween current high
+     * water mark and log end i.e. if high watermark is updated consumer will
+     * receive data. This way we will only update high watermark when leadership
+     * changed after partition is moved or redpanda restarted.
+     */
+    const auto log_end = raft::details::next_offset(
+      _translator->from_log_offset(_partition->dirty_offset()));
+
+    if (fetch_offset > high_watermark() && fetch_offset <= log_end) {
+        while (_partition->high_watermark() <= _partition->dirty_offset()) {
```

Comment:
can you add some clarity to how this works?

`log_end` is a function of `dirty_offset` (computed above), and then we have two nested conditions where the outer is based on `log_end` but the inner is based on the raw `dirty_offset`. further, the outer uses `high_watermark()` and the inner uses `partition->high_watermark()`. if this is all domain knowledge that should be known from context then it seems ok, but I found this confusing.